apps/passport-client/src/validateState.ts

```diff
@@ -0,0 +1,158 @@
+ * determines the user is logged in or the {@link forceCheckPCDs} argument is `true`.
+ */
+function validateAppState(
+  self: User | undefined,
+  identity: Identity | undefined,
+  pcds: PCDCollection | undefined,
+  forceCheckPCDs?: boolean
+): ValidationErrors {
+  const validationErrors: string[] = [];
+
+  const loggedOut = !self;
+
+  const identityPCDFromCollection = pcds?.getPCDsByType(
+    SemaphoreIdentityPCDPackage.name
+  )?.[0] as SemaphoreIdentityPCD | undefined;
+
+  if (forceCheckPCDs || !loggedOut) {
+    if (!identityPCDFromCollection) {
+      validationErrors.push(
+        "'pcds' field in app state does not contain an identity PCD"
+      );
+    }
+  }
+
+  if (loggedOut) {
+    return {
+      errors: validationErrors,
+      userUUID: undefined
+    };
+  }
+
+  if (!self) {
+    validationErrors.push("missing 'self'");
+  }
+
+  if (!identity) {
+    validationErrors.push("missing 'identity'");
+  }
+
+  if (!pcds) {
+    validationErrors.push("missing 'pcds'");
+  }
+
+  if (pcds?.size() === 0) {
+    validationErrors.push("'pcds' contains no pcds");
+  }
+
+  const identityFromPCDCollection = identityPCDFromCollection?.claim?.identity;
+  const commitmentOfIdentityPCDInCollection =
+    identityFromPCDCollection?.commitment?.toString();
+  const commitmentFromSelfField = self?.commitment;
+  const commitmentFromIdentityField = identity?.commitment?.toString();
+
+  if (commitmentOfIdentityPCDInCollection !== commitmentFromSelfField) {
+    validationErrors.push(
+      `commitment of identity pcd in collection (${commitmentOfIdentityPCDInCollection})` +
+        ` does not match commitment in 'self' field of app state (${commitmentFromSelfField})`
+    );
+  }
+
+  if (commitmentFromSelfField !== commitmentFromIdentityField) {
+    validationErrors.push(
+      `commitment in 'self' field of app state (${commitmentFromSelfField})` +
+        ` does not match commitment of 'identity' field of app state (${commitmentFromIdentityField})`
+    );
+  }
+
+  if (commitmentFromIdentityField !== commitmentOfIdentityPCDInCollection) {
+    validationErrors.push(
+      `commitment of 'identity' field of app state (${commitmentFromIdentityField})` +
+        ` does not match commitment of identity pcd in collection (${commitmentOfIdentityPCDInCollection})`
+    );
+  }
+
+  return {
+    errors: validationErrors,
+    userUUID: self?.uuid
+  };
+}
+
+/**
+ * Logs validation errors to the console, and uploads them to the server so that
+ * we have records and are able to identify common types of errors. Does not leak
+ * sensitive information, such as decrypted versions of e2ee storage.
+ */
+async function logValidationErrors(errors: ValidationErrors): Promise<void> {
+  try {
+    const user = loadSelf();
+    errors.userUUID = errors.userUUID ?? user?.uuid;
+    console.log(`encountered state validation errors: `, errors);
+    await requestLogToServer(appConfig.zupassServer, "state-validation-error", {
```

Comment:
I expected to see the userUUID field at the top level here, rather than inside of the errors object.  The only reason I think I'd care is if having it deeper down makes it harder to filter/group by this field in server-side logs.  Do you have insight about that?

Reply:
I'm not sure it would make it harder to query the logs, but it I am not sure how much more difficult it would make honeycomb tracing, given I've forgotten whether it flattens objects. I'll unwrap the fields and think about that later. https://github.com/proofcarryingdata/zupass/issues/1365

Reply:
Not applied yet?

Reply:
Not yet; I'll resolve the comment once it's done